### PR TITLE
Do not capture buttons in AccountsList that belong to inner VerticalLayout widget

### PR DESCRIPTION
**Pull request by mymedia2**
This fixes a use-after-free error (double destruction) in the main menu right after account switching. The Ui::VerticalLayout::add method seems to return a weak pointer that should not be given to base::unique_qptr or anything.

The regression was introduced in commit f31be7784b9ef0790408a353e3d53e6022553d1f.

```
==121545== Invalid read of size 8
==121545==    at 0x603F7D: QWeakPointer<QObject>::internalData() const (qsharedpointer_impl.h:698)
==121545==    by 0x603FE8: QPointer<QObject>::data() const (qpointer.h:77)
==121545==    by 0xA2DCC6: object_ptr<Ui::RpWidget>::data() const (object_ptr.h:68)
==121545==    by 0xA3284A: object_ptr<Ui::RpWidget>::operator Ui::RpWidget*() const (object_ptr.h:71)
==121545==    by 0x6B70AEC: Ui::VerticalLayout::childHeightUpdated(Ui::RpWidget*)::{lambda(Ui::VerticalLayout::Row const&)#1}::operator()(Ui::VerticalLayout::Row const&) const (vertical_layout.cpp:144)
...
```

I'm attaching the full Valgrind log. You can see there errors about uninitialized values. I've fixed some of them in other pull requests, desktop-app/lib_rpl#2, desktop-app/lib_ui#86, [tdesktop.log](https://github.com/telegramdesktop/tdesktop/files/8454352/tdesktop.log).

### CODE REVIEW

File: Telegram/SourceFiles/settings/settings_information.cpp [hunk 1]
```diff
@@ -78,9 +78,7 @@ class AccountsList final {
 	int _outerIndex = 0;
 
 	Ui::SlideWrap<Ui::SettingsButton> *_addAccount = nullptr;
-	base::flat_map<
-		not_null<Main::Account*>,
-		base::unique_qptr<Ui::SettingsButton>> _watched;
+	base::flat_map<not_null<Main::Account*>, Ui::SettingsButton*> _watched;
```
Comment (23rd) [inline Telegram/SourceFiles/settings/settings_information.cpp:81]:
```suggestion
	base::flat_map<
		not_null<Main::Account*>,
		not_null<Ui::SettingsButton*>> _watched;
```

File: Telegram/SourceFiles/settings/settings_information.cpp [hunk 2]
```diff
@@ -730,7 +728,7 @@ void AccountsList::rebuild() {
 				order.reserve(inner->count());
 				for (auto i = 0; i < inner->count(); i++) {
 					for (const auto &[account, button] : _watched) {
-						if (button.get() == inner->widgetAt(i)) {
+						if (button == inner->widgetAt(i)) {
```
Comment (23rd) [inline Telegram/SourceFiles/settings/settings_information.cpp:731]:
```suggestion
						if (button.get() == inner->widgetAt(i)) {
```

File: Telegram/SourceFiles/settings/settings_information.cpp (hunk 1 above)
Comment (mymedia2) [reply Telegram/SourceFiles/settings/settings_information.cpp:81]: @23rd Value of this map may set nullish at line 656 or 751.

Comment (23rd) [reply Telegram/SourceFiles/settings/settings_information.cpp:81]: Yes, I had updated my suggestion.

**Comment by john-preston** [comment]
@mymedia2 Are you sure this helps? base::unique_qptr holds QPointer inside 🤔

**Comment by mymedia2** [comment]
@john-preston as I can see, base:unique_qptr calls to the destructor of pointed object. That's not appropriate since these Ui::SettingsButtons get destroyed by its container, Ui::VerticalLayout. QPointer has no such logic, it safely becomes NULL when the object is destroyed.

**Comment by john-preston** [comment]
@mymedia2 fixed by https://github.com/desktop-app/lib_ui/commit/aa155db0bd11485643511da210fa7629b48347e8

**Pull request closed without merging**